File: tests/examples/deduction.cpp

#include <stdint.h>

#include <array>  // std::array
#include <map>

template <typename A, typename B, typename C>
class AutoClass {
    std::map<A, std::pair<B, C>> t;
    static auto key   = std::map<A, std::pair<B, C>>::key_type();
    static auto value = std::map<A, std::pair<B, C>>::value_type();
};

class DecltypeClass {
    const decltype(1) a = 1;
};

template <typename T>
class DependentTypeClass {
    T a;
    decltype(a) b;
};

File: tests/examples/deduction.py
from examples.utils import *

def case_deduction():
    data = read_json('deduction.cpp.json')
    expect_true(any((f['type'] == 'auto' for f in data['AutoClass']['fields'])))
    expect_eq(data['DecltypeClass']['fields'][0]['types'][0]['type'] , 'int')
    expect_true(any((f['name'] == 'a' and f['types'][0]['type'] == 'T' for f in data['DependentTypeClass']['fields'])))
    expect_true(any((f['name'] == 'b' and f['types'][0]['type'] == '<dependent type>' for f in data['DependentTypeClass']['fields'])))
